File: agora-android/app/src/main/java/com/kobildn/agora/Bridge.kt
package com.kobildn.agora

import android.os.Handler
import android.os.Looper

/**
 * The relay engine — Kotlin port of the extension's background.js / the
 * desktop app's lib/bridge.js. Same behaviour: forwarding with turn delay,
 * sender labels + one-time intro, mutual [STOP_BRIDGE] detection, voice
 * commands, dedupe, turn caps, single-target interject routing.
 *
 * Host responsibilities (MainActivity):
 *  - sendToSite(site, text): inject text into that site's WebView, returns
 *    false if the panel isn't ready
 *  - onStateChanged(): re-render UI (log, bridge button)
 */
class Bridge(
    private val sendToSite: (String, String) -> Boolean,
    private val onStateChanged: () -> Unit
) {
    data class LogEntry(val sender: String, val message: String, val timestamp: Long)

    data class Settings(
        var turnDelay: Int = 3,
        var maxTurns: Int = 0,
        var labelMessages: Boolean = true,
        var interjectTarget: String = "Claude"
    )

    val settings = Settings()
    val log = ArrayList<LogEntry>()
    var bridgeActive = false
        private set

    private var turnCount = 0
    private val introSentTo = hashMapOf("DeepSeek" to false, "Claude" to false)
    private var lastStopSender: String? = null
    private var lastStopTime = 0L

    private val main = Handler(Looper.getMainLooper())

    companion object {
        const val STOP_TOKEN = "[STOP_BRIDGE]"
        const val STOP_WINDOW_MS = 3 * 60 * 1000L
        const val DEDUPE_WINDOW_MS = 60 * 1000L
        private val PAUSE_WORDS = setOf("stop", "pause", "halt")
        private val RESUME_WORDS = setOf("resume", "continue", "start")
    }

    private fun otherSite(name: String): String? = when (name) {
        "DeepSeek" -> "Claude"
        "Claude" -> "DeepSeek"
        else -> null
    }

    private fun detectCommand(text: String): String? {
        val words = text.lowercase().replace(Regex("[^\\w\\s]"), "").split(Regex("\\s+")).filter { it.isNotEmpty() }
        if (words.isEmpty() || words.size > 4) return null
        if (words.any { it in PAUSE_WORDS }) return "pause"
        if (words.any { it in RESUME_WORDS }) return "resume"
        return null
    }

    // Token must be the literal trailing content of the message — mentions
    // elsewhere never trigger (extension issue #18, three rounds of this).
    private fun endsWithStopToken(text: String): Boolean = text.trim().endsWith(STOP_TOKEN)

    fun addLogEntry(sender: String, message: String) {
        log.add(LogEntry(sender, message, System.currentTimeMillis()))
        if (log.size > 200) log.removeAt(0)
        onStateChanged()
    }

    fun toggleBridge() {
        bridgeActive = !bridgeActive
        turnCount = 0
        introSentTo["DeepSeek"] = false
        introSentTo["Claude"] = false
        lastStopSender = null
        onStateChanged()
    }

    fun clearLog() {
        log.clear()
        turnCount = 0
        onStateChanged()
    }

    private fun labelText(from: String, target: String, text: String): String {
        if (!settings.labelMessages) return text
        var intro = ""
        if (introSentTo[target] == false) {
            intro = "[Bridge notice: You are in a relayed conversation with another AI, $from. " +
                "Messages prefixed [$from] are written by that AI, not by a human. " +
                "A human moderator supervises and may interject; their messages are prefixed [Human]. " +
                "When you and the other AI have both independently reached a genuine natural " +
                "conclusion — not just a lull, and not proactively suggesting it end — end your " +
                "ENTIRE message with the literal token $STOP_TOKEN as the very last characters, " +
                "with nothing else after it. Merely mentioning the token elsewhere will NOT " +
                "trigger anything — only trailing use does. If you and the other AI are just " +
                "repeating \"agreed\" / \"nothing more to add\" with no new content, that " +
                "repetition IS the natural conclusion — use the token right then.]\n\n"
            introSentTo[target] = true
        }
        return "$intro[$from]: $text"
    }

    private fun checkMutualStop(sender: String, message: String): Boolean {
        if (!endsWithStopToken(message)) return false
        val now = System.currentTimeMillis()
        val prev = lastStopSender
        if (prev != null && prev != sender && (now - lastStopTime) < STOP_WINDOW_MS) {
            bridgeActive = false
            lastStopSender = null
            addLogEntry("System", "Bridge stopped — both AIs signaled agreement.")
            return true
        }
        lastStopSender = sender
        lastStopTime = now
        return false
    }

    /** An AI response captured by the injected site script. */
    fun handleNewMessage(sender: String, message: String) {
        val cutoff = System.currentTimeMillis() - DEDUPE_WINDOW_MS
        val duplicate = log.any { it.sender == sender && it.timestamp > cutoff && it.message == message }
        if (duplicate) return

        addLogEntry(sender, message)

        if (checkMutualStop(sender, message)) return
        if (!bridgeActive) return

        if (settings.maxTurns > 0 && turnCount >= settings.maxTurns) {
            bridgeActive = false
            addLogEntry("System", "Bridge paused after ${settings.maxTurns} turn${if (settings.maxTurns == 1) "" else "s"}.")
            return
        }

        val delayMs = (settings.turnDelay.coerceAtLeast(0)) * 1000L
        main.postDelayed({ forwardMessage(sender, message) }, delayMs)
    }

    private fun forwardMessage(sender: String, message: String) {
        if (!bridgeActive) return
        val target = otherSite(sender) ?: return
        turnCount++
        val delivered = sendToSite(target, labelText(sender, target, message))
        if (!delivered) {
            addLogEntry("System", "⚠️ Could not deliver to the $target panel — is it loaded?")
        }
        onStateChanged()
    }

    /**
     * User interjection → deliver to the chosen target only; the other AI
     * hears it through the normal relay path. Broadcasting to both spawns two
     * parallel reply chains (diagnosed by the bridged AIs themselves).
     */
    fun handleUserMessage(text: String, target: String) {
        addLogEntry("User", text)

        when (detectCommand(text)) {
            "pause" -> {
                bridgeActive = false
                addLogEntry("System", "Bridge paused by voice command.")
            }
            "resume" -> {
                bridgeActive = true
                turnCount = 0
                addLogEntry("System", "Bridge resumed by voice command.")
            }
        }

        val outgoing = if (settings.labelMessages) "[Human]: $text" else text
        val command = detectCommand(text)
        val resolved = if (command != null) "Both" else target
        val targets = if (resolved == "Both") listOf("DeepSeek", "Claude") else listOf(resolved)
        for (name in targets) sendToSite(name, outgoing)
    }

    /** Manual forward: hand the last captured response from one AI to the other. */
    fun forwardLast(from: String): String? {
        val entry = log.lastOrNull { it.sender == from } ?: return "No $from response found to forward."
        val target = otherSite(from) ?: return "Unknown source."
        val delivered = sendToSite(target, labelText(from, target, entry.message))
        if (!delivered) return "Could not deliver to the $target panel — is it loaded?"
        addLogEntry("System", "Manually forwarded last $from response to $target.")
        return null
    }
}

File: agora-app/lib/bridge.js
// The relay engine — a direct port of the extension's background.js, minus
// every MV3 workaround it needed (chrome.alarms backstops, persisted pending
// forwards, storage-read-per-event). An Electron main process is a normal
// long-running Node process, so plain setTimeout and in-memory state are
// reliable here.

const { ipcMain } = require('electron');

const STOP_TOKEN = '[STOP_BRIDGE]';
const STOP_WINDOW_MS = 3 * 60 * 1000;
const DEDUPE_WINDOW_MS = 60 * 1000;
const SHARE_LOG_MAX_CHARS = 15000;

const PAUSE_WORDS = ['stop', 'pause', 'halt'];
const RESUME_WORDS = ['resume', 'continue', 'start'];

function detectCommand(text) {
  const words = text.toLowerCase().replace(/[^\w\s]/g, '').split(/\s+/).filter(Boolean);
  if (words.length === 0 || words.length > 4) return null;
  if (words.some(w => PAUSE_WORDS.includes(w))) return 'pause';
  if (words.some(w => RESUME_WORDS.includes(w))) return 'resume';
  return null;
}

// Token must be the literal trailing content of the message — mentions
// elsewhere never trigger (see extension issue #18's three rounds of this).
function endsWithStopToken(text) {
  return text.trim().endsWith(STOP_TOKEN);
}

class Bridge {
  /**
   * @param store lib/store.js Store instance
   * @param sendToSite (siteName, text) => boolean — inject text into a site view
   * @param otherSite (siteName) => siteName
   */
  constructor(store, sendToSite, otherSite) {
    this.store = store;
    this.sendToSite = sendToSite;
    this.otherSite = otherSite;
    this.wireIpc();
  }

  wireIpc() {
    ipcMain.on('site:newMessage', (e, { sender, message }) => {
      this.handleNewMessage(sender, message);
    });

    ipcMain.on('site:selectorError', (e, { site, what }) => {
      this.store.addLogEntry('System', `⚠️ Could not find ${what} on ${site} — its UI may have changed.`);
    });

    ipcMain.on('console:toggleBridge', () => {
      const { bridgeActive } = this.store.get();
      this.store.set({
        bridgeActive: !bridgeActive,
        turnCount: 0,
        introSentTo: { DeepSeek: false, Claude: false },
        lastStop: null
      });
    });

    ipcMain.on('console:interject', (e, { message, target }) => {
      this.handleUserMessage(message, target);
    });

    ipcMain.handle('console:forwardLast', (e, from) => this.forwardLastResponse(from));
    ipcMain.handle('console:shareLog', (e, target) => this.shareLog(target));

    ipcMain.on('console:setSettings', (e, settings) => {
      this.store.set({ settings: { ...this.store.get().settings, ...settings } });
    });

    ipcMain.on('console:clearLog', () => {
      this.store.set({ conversationLog: [], turnCount: 0 });
    });
  }

  labelText(from, target, text) {
    const state = this.store.get();
    if (!state.settings.labelMessages) return text;

    let intro = '';
    if (!state.introSentTo[target]) {
      intro = `[Bridge notice: You are in a relayed conversation with another AI, ${from}. ` +
              `Messages prefixed [${from}] are written by that AI, not by a human. ` +
              `A human moderator supervises and may interject; their messages are prefixed [Human]. ` +
              `When you and the other AI have both independently reached a genuine natural ` +
              `conclusion — not just a lull, and not proactively suggesting it end — end your ` +
              `ENTIRE message with the literal token ${STOP_TOKEN} as the very last characters, ` +
              `with nothing else after it. Merely mentioning the token elsewhere in a message ` +
              `will NOT trigger anything — only trailing use does. IMPORTANT: if you notice you ` +
              `and the other AI are just repeating "agreed" / "nothing more to add" back and ` +
              `forth with no new content, that repetition IS the natural conclusion — use the ` +
              `token right then instead of continuing to trade acknowledgments.]\n\n`;
      this.store.set({ introSentTo: { ...state.introSentTo, [target]: true } });
    }
    return `${intro}[${from}]: ${text}`;
  }

  checkMutualStop(sender, message) {
    if (!endsWithStopToken(message)) return false;

    const { lastStop } = this.store.get();
    const now = Date.now();

    if (lastStop && lastStop.sender !== sender && (now - lastStop.time) < STOP_WINDOW_MS) {
      this.store.set({ bridgeActive: false, lastStop: null });
      this.store.addLogEntry('System', 'Bridge stopped — both AIs signaled agreement.');
      return true;
    }
    this.store.set({ lastStop: { sender, time: now } });
    return false;
  }

  handleNewMessage(sender, message) {
    // Dedupe: the observer can capture the same message from two DOM nodes
    const cutoff = Date.now() - DEDUPE_WINDOW_MS;
    const isDuplicate = this.store.get().conversationLog.some(e =>
      e.sender === sender && e.timestamp > cutoff && e.message === message
    );
    if (isDuplicate) return;

    this.store.addLogEntry(sender, message);

    if (this.checkMutualStop(sender, message)) return;

    const state = this.store.get();
    if (!state.bridgeActive) return;

    if (state.settings.maxTurns > 0 && state.turnCount >= state.settings.maxTurns) {
      this.store.set({ bridgeActive: false });
      this.store.addLogEntry('System', `Bridge paused after ${state.settings.maxTurns} turn${state.settings.maxTurns === 1 ? '' : 's'}.`);
      return;
    }

    const delayMs = (state.settings.turnDelay || 0) * 1000;
    setTimeout(() => this.forwardMessage(sender, message), delayMs);
  }

  forwardMessage(sender, message) {
    const state = this.store.get();
    if (!state.bridgeActive) return;

    const target = this.otherSite(sender);
    if (!target) return;

    this.store.set({ turnCount: state.turnCount + 1 });
    const delivered = this.sendToSite(target, this.labelText(sender, target, message));
    if (!delivered) {
      this.store.addLogEntry('System', `⚠️ Could not deliver to the ${target} panel — is it loaded?`);
    }
  }

  // User interjection → deliver to the chosen target; the other AI hears it
  // through the normal relay path (broadcasting to both spawns two parallel
  // reply chains — diagnosed by the bridged AIs themselves, extension #17)
  handleUserMessage(text, target) {
    this.store.addLogEntry('User', text);

    const command = detectCommand(text);
    if (command === 'pause') {
      this.store.set({ bridgeActive: false });
      this.store.addLogEntry('System', 'Bridge paused by voice command.');
    } else if (command === 'resume') {
      this.store.set({ bridgeActive: true, turnCount: 0 });
      this.store.addLogEntry('System', 'Bridge resumed by voice command.');
    }

    const state = this.store.get();
    const outgoing = state.settings.labelMessages ? `[Human]: ${text}` : text;

    const resolved = command ? 'Both' : (target || state.settings.interjectTarget || 'Claude');
    const targets = resolved === 'Both' ? ['DeepSeek', 'Claude'] : [resolved];
    for (const name of targets) {
      this.sendToSite(name, outgoing);
    }
  }

  // Manual forward: hand the last captured response from one AI to the other.
  // Works with the bridge off. The message is already in the log (capture
  // always logs), so only a System note is added.
  forwardLastResponse(from) {
    const log = this.store.get().conversationLog;
    const entry = [...log].reverse().find(e => e.sender === from);
    if (!entry) {
      return { success: false, error: `No ${from} response found to forward.` };
    }

    const target = this.otherSite(from);
    const delivered = this.sendToSite(target, this.labelText(from, target, entry.message));
    if (!delivered) {
      return { success: false, error: `Could not deliver to the ${target} panel — is it loaded?` };
    }
    this.store.addLogEntry('System', `Manually forwarded last ${from} response to ${target}.`);
    return { success: true };
  }

  shareLog(target) {
    const log = this.store.get().conversationLog;
    if (!log.length) return { success: false, error: 'Log is empty — nothing to share.' };

    let body = log
      .map(e => `[${new Date(e.timestamp).toLocaleTimeString()}] ${e.sender}: ${e.message}`)
      .join('\n\n');
    if (body.length > SHARE_LOG_MAX_CHARS) {
      body = '…(earlier messages truncated)…\n\n' + body.slice(-SHARE_LOG_MAX_CHARS);
    }

    const outgoing =
      `[Human moderator: sharing the bridge's conversation log below for your reference — ` +
      `this is the app's own record (with attribution and timestamps), not a new ` +
      `conversational turn. Acknowledge briefly; no need to re-analyze every entry.]\n\n${body}`;

    const resolved = target || 'Both';
    const targets = resolved === 'Both' ? ['DeepSeek', 'Claude'] : [resolved];
    const delivered = targets.filter(name => this.sendToSite(name, outgoing));

    if (!delivered.length) {
      return { success: false, error: 'No reachable AI panel.' };
    }
    this.store.addLogEntry('System', `Log shared with ${delivered.join(' and ')} (${log.length} messages).`);
    return { success: true };
  }
}

module.exports = { Bridge, STOP_TOKEN };
